src/array_api_extra/_lib/_at.py

```diff
@@ -355,9 +370,70 @@ def _op(
+        try:  # We first try to use the backend's __setitem__ if available
             x[idx] = y
-        return x
+            return x
+        except IndexError as e:
+            if "Fancy indexing" not in str(e):  # Avoid masking other index errors
+                raise e
+            # Work around lack of fancy indexing __setitem__
+            if (
+                is_array_api_obj(idx)
+                and xp.isdtype(idx.dtype, "integral")
+                and idx.ndim == 1
+            ):
+                # Vectorize the operation using boolean indexing
+                # For non-unique indices, take the last occurrence. This requires
+                # masks for x and y that create matching shapes.
+                # We first create the mask for x
+                u_idx, _ = xp.unique_inverse(idx)
+                # Convert negative indices to positive, otherwise they won't get matched
+                u_idx_pos = xp.where(u_idx < 0, x.shape[0] + u_idx, u_idx)
+                # Check for out of bounds indices
+                oob_index = None
+                if is_lazy_array(u_idx_pos):
+                    pass  # Lazy arrays cannot check for out of bounds indices
+                elif xp.any(pos_oob := u_idx_pos >= x.shape[0]):
+                    first_oob_idx = xp.argmax(xp.astype(pos_oob, xp.int32))
+                    oob_index = int(u_idx[first_oob_idx])
+                elif xp.any(neg_oob := u_idx_pos < 0):
+                    first_oob_idx = xp.argmax(xp.astype(neg_oob, xp.int32))
+                    oob_index = int(u_idx[first_oob_idx])
+                if oob_index is not None:
+                    msg = (
+                        f"index {oob_index} is out of bounds for array of "
+                        f"shape {x.shape}"
+                    )
+                    raise IndexError(msg) from e
+
+                x_mask = xp.any(xp.arange(x.shape[0])[..., None] == u_idx_pos, axis=-1)
+                # If y is a scalar or 0D, we are done
+                if not is_array_api_obj(y) or y.ndim == 0:
+                    x[x_mask] = y
+                    return x
+                if y.shape[0] != idx.shape[0]:
+                    msg = (
+                        f"shape mismatch: value array of shape {y.shape} could not be "
+                        f"broadcast to indexing result of shape {idx.shape}"
+                    )
+                    raise ValueError(msg) from e
+                # If not, create a mask for y. Get last occurrence of each unique index
+                cmp = u_idx_pos[:, None] == u_idx_pos[None, :]
+                # Ignore later matches
+                lower_tri_mask = (
+                    xp.arange(y.shape[0])[:, None] >= xp.arange(y.shape[0])[None, :]
+                )
+                masked_cmp = cmp & lower_tri_mask
+                # For each position i, count how many matches occurred before i
+                prior_matches = xp.sum(xp.astype(masked_cmp, xp.int32), axis=-1)
```

Comment:
philosophical nit: automatically use native int
```suggestion
                prior_matches = xp.sum(xp.where(masked_cmp, 1, 0), axis=-1)
```

Reply:
Not valid in array-api-strict:

> TypeError: Two scalars not allowed, got type(x1) = <class 'int'> and type(x2) =<class 'int'>